Library/Homebrew/extend/pathname.rb

```diff
@@ -391,9 +394,37 @@ def env_script_all_files(dst, env)
     Pathname.glob("#{self}/*") do |file|
       next if file.directory?
 
-      dst.install(file)
-      new_file = dst.join(file.basename)
-      file.write_env_script(new_file, env)
+      # things like meson get mighty confused by the env scripts because
+      # it obscures the original binary name, which meson uses to calculate
+      # internal python paths. So if the target is already using a #!, we
+      # simply enhance it rather than creating the env script.
+      #
+      # The usage of -S to env is the only somewhat exotic thing here. It's
+      # in all env binaries that homebrew runs on.
+
+      used_env = false
+
+      if file.read(2) == "#!"
+        lines = file.readlines
+
+        unless lines.empty?
+          first = lines[0]
+          if first
+            env_export = +""
+            env.each { |key, value| env_export << "#{key}=\"#{value}\" " }
+            lines[0] = "#!/usr/bin/env -S #{env_export}#{first[2..]}"
```

Comment:
This will break env scripts that do something like
```ruby
bin.env_script_all_files libexec, FOO: "bar", PATH: "/path/to/foo/bin:$PATH"
```
because `env` won't interpolate the `$PATH`.

Reply:
That will work fine, `env`'s `-S` supports interpolation. (Example from linux because the man page doesn't say so https://github.com/coreutils/coreutils/blob/master/src/env.c#L450, the macos man page says that the interpolation is supported).

Reply:
No, It will **not** work fine. 
```
❯ cat foo
#!/usr/bin/env -S "FOO=bar:$FOO" zsh

echo $FOO
❯ ./foo
env: Only ${VARNAME} expansion is supported, error at: $FOO"
❯ FOO=baz ./foo
env: Only ${VARNAME} expansion is supported, error at: $FOO"
```

 Which version of `env` did you test this with? This isn't even specific to macOS `env`.
```
❯ cat foo
#!/usr/local/opt/coreutils/libexec/gnubin/env -S "FOO=bar:$FOO" zsh

echo $FOO
❯ ./foo
env: only ${VARNAME} expansion is supported, error at: $FOO"
```

It only works if the variable to be interpolated is protected by braces.
```
❯ cat foo
#!/usr/bin/env -S "FOO=bar:${FOO}" zsh

echo $FOO
❯ FOO=baz ./foo
bar:baz
```

However, there are a lot of instances of variable interpolation into env scripts that are _not_ protected by braces, so this change will break all of those and is likely to cause issues every time someone writes an env script (because who knows if they'll even know about or remember to add the `{}`?).

If we want to do this, we'd need to, at minimum, add an audit that makes sure that variable interpolation is protected by braces. That said, it's still not clear why we want to mess with the shebang when we could just do `exec -a`.

Reply:
Ah geez, you're totally right. Thank you so much for checking, you're totally right the braces are needed.

I didn't test interpolation, only consulted (and misunderstood) the docs.

You're right this could just be too much trouble.

As for the usage with `exec -a`, that doesn't work the target also uses a shebang because the kernel replaces argv[0] to be the real path of the target, regardless of what the value of argue[0] was when the process was executed.

Reply:
> As for the usage with `exec -a`, that doesn't work the target also uses a shebang because the kernel replaces argv[0] to be the real path of the target, regardless of what the value of argue[0] was when the process was executed.

In that case the solution for `write_env_script` isn't really the correct one either.